### PR TITLE
Tensor Ring implementation

### PR DESCRIPTION
**Pull request by merajhashemi**
This PR implements Tensor Ring decomposition using recursive SVD; see [Algorithm 1](https://arxiv.org/abs/1606.05535).

### CODE REVIEW

File: tensorly/decomposition/_tr.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    factors = [None] * n_dim
+
+    # Getting the first factor
+    unfolding = tl.reshape(input_tensor, (tensor_size[0], -1))
+
+    n_row, n_column = unfolding.shape
+    # TODO: Proper warning message
+    if rank[0] * rank[1] > min(n_row, n_column):
+        raise ValueError(f'rank[{mode}] * rank[{mode + 1}] = {rank[0] * rank[1]} is larger than'
+                         'first matricization dimensions.\n'
```
Comment (JeanKossaifi) [inline tensorly/decomposition/_tr.py:51]: Would be good to have a clearer message, e.g.
The mismatch between rank[mode] and rank[0],

The best is to have a message that points precisely the error to the users and helps to fix it. 
We could for instance say the first unfolding is of size (x, y) so will have rank at most min(x, y) and yet got rank[0]*rank[1]=z > min(x, y).

File: tensorly/decomposition/_tr.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        Decomposes `input_tensor` into a sequence of order-3 tensors (factors) [1]_.
+
+    Parameters
+    ----------
+    input_tensor : tensorly.tensor
+    rank : {int, int list}
+            maximum allowable TR rank of the factors
+            if int, then this is the same for all the factors
+            if int list, then rank[k] is the rank of the kth factor
+    mode : int, default is 0
```
Comment (JeanKossaifi) [inline tensorly/decomposition/_tr.py:17]: When would we want to not start by 0? Does it change much the final decomposition?

File: tensorly/tr_tensor.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        rank_prev, _, rank_next = factor.shape
+        factor = tl.reshape(factor, (rank_prev, -1))
+        full_tensor = tl.dot(full_tensor, factor)
+        full_tensor = tl.reshape(full_tensor, (-1, rank_next))
+
+    full_tensor = tl.reshape(full_tensor, (factors[-1].shape[2], -1, factors[-1].shape[0]))
+    full_tensor = tl.moveaxis(full_tensor, 0, -1)
+    full_tensor = tl.reshape(full_tensor, (-1, factors[-1].shape[0] * factors[-1].shape[2]))
+    factor = tl.moveaxis(factors[-1], -1, 1)
+    factor = tl.reshape(factor, (-1, full_shape[-1]))
```
Comment (JeanKossaifi) [inline tensorly/tr_tensor.py:41]: That seems like a lot of reshapes and moveaxis, though I haven't had the time to look if we can avoid some.

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @merajhashemi, we've been meaning to add tensor ring for a while!

Looks good overall, I would just make a few changes in the decomposition function and add some unit-tests!

**Comment by JeanKossaifi** [comment]
@merajhashemi are you still working on this - that's a great feature to have!

**Comment by JeanKossaifi** [comment]
@merajhashemi looks good to me - I had just left some comments in the code, if you can have a quick look and make any change where appropriate then I think we're good to merge.

File: tensorly/decomposition/_tr.py (hunk 2 above)
Comment (merajhashemi) [reply tensorly/decomposition/_tr.py:17]: The authors in the original paper state:
> TR-SVD might obtain different representations by choosing a different mode as the start point. This indicates that TR-ranks is not necessary to be the global optimum in TR-SVD.

Generally, the ranks on the sides are smaller than the ones in middle. Changing the mode allows to shift the sides and to explore different ranks.

File: tensorly/tr_tensor.py (hunk 3 above)
Comment (merajhashemi) [reply tensorly/tr_tensor.py:41]: I think only line 35 is redundant for the last factor in the for loop, as it is reshaped again in line 37.

**Comment by JeanKossaifi** [comment]
Awesome, thanks @merajhashemi, looks great, merging!

**Pull request merged**